solidity/src/processor/libs/ProcessorMessage.sol

```diff
@@ -0,0 +1,367 @@
+        Atomic,
+        NonAtomic
+    }
+
+    /// @notice Container for either atomic or non-atomic subroutine
+    /// @param subroutineType Type of contained subroutine
+    /// @param atomic Atomic subroutine if type is Atomic
+    /// @param nonAtomic Non-atomic subroutine if type is NonAtomic
+    struct Subroutine {
+        SubroutineType subroutineType;
+        AtomicSubroutine atomic;
+        NonAtomicSubroutine nonAtomic;
+    }
+
+    /// @notice Result of parsing messages from batch
+    /// @param messages Array of parsed message bytes
+    /// @param offset Next position in byte array after messages
+    struct MessagesResult {
+        bytes[] messages;
+        uint256 offset;
+    }
+
+    /// @notice Get type of authorization message that we are executing on the processor
+    /// @param message Raw message bytes
+    /// @return bytes1 Authorization Message Type value (0x00 for EnqueueMsgs, 0x01 for EvictMsgs, 0x02 for InsertMsgs, 0x03 for Pause and 0x04 for Resume)
+    function authorizationMessageType(bytes calldata message) internal pure returns (bytes1) {
+        return message[0];
+    }
+
+    /// @notice Get queue position for the message batch in the queue (Only used for EvictMsgs and InsertMsgs, for EnqueueMsgs the value will be 0)
+    /// @param message Raw message bytes
+    /// @return uint64 Queue Position
+    function queuePosition(bytes calldata message) internal pure returns (uint64) {
+        bytes memory slice = message[1:9];
+        return uint64(bytes8(slice));
+    }
+
+    /// @notice Get priority from message batch
+    /// @param message Raw message bytes
+    /// @return bytes1 Priority value (0x00 for Medium, 0x01 for High)
+    function priority(bytes calldata message) internal pure returns (bytes1) {
+        return message[9];
+    }
+
+    /// @notice Get execution ID from message batch
+    /// @param message Raw message bytes
+    /// @return uint64 Execution ID
+    function id(bytes calldata message) internal pure returns (uint64) {
+        bytes memory slice = message[10:18];
+        return uint64(bytes8(slice));
+    }
+
+    /// @notice Get number of messages in batch
+    /// @param message Raw message bytes
+    /// @return uint64 Number of messages
+    function numMessages(bytes calldata message) internal pure returns (uint64) {
+        bytes memory slice = message[18:26];
+        return uint64(bytes8(slice));
+    }
+
+    /// @notice Parse messages from batch
+    /// @param message Raw message bytes
+    /// @return MessagesResult Parsed messages and next offset
+    function messages(bytes calldata message) internal pure returns (MessagesResult memory) {
+        uint64 count = numMessages(message);
+        bytes[] memory result = new bytes[](count);
+
+        uint256 offset = 26; // Start after header (1 + 8 + 1 + 8 + 8 bytes)
+
+        for (uint64 i = 0; i < count; i++) {
+            uint64 length = uint64(bytes8(message[offset:offset + 8]));
```

Comment:
should we validate that `offset:offset + 8` is a valid index in the message?

Reply:
Will be reworked, but in theory no validation would be needed because we are doing the encoding as well and would ensure it would be a valid index